The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3196850995/jobs/5219318867) [(plain)](https://github.com/rust-lang/rust/commit/84ab71266404d867964a8cd7bf1fc1b3983a303c/checks/8742082019/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: variable does not need to be mutable
  --> library/core/src/future/poll_fn.rs:62:13
   |
62 |     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<T> {
   |             |
   |             help: remove this `mut`
   |
   |
   = note: `-D unused-mut` implied by `-D warnings`

error[E0596]: cannot borrow data in dereference of `Pin<&mut PollFn<F>>` as mutable
   |
   |
63 |         (&mut self.f)(cx)
   |         ^^^^^^^^^^^^^ cannot borrow as mutable
   |
   = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Pin<&mut PollFn<F>>`
For more information about this error, try `rustc --explain E0596`.
error: could not compile `core` due to 2 previous errors
Build completed unsuccessfully in 0:01:37
```

</details>

--- a/library/core/src/future/poll_fn.rs
+++ b/library/core/src/future/poll_fn.rs
@@ -5,7 +5,9 @@ use crate::task::{Context, Poll};
 
 /// Creates a future that wraps a function returning [`Poll`].
 ///
-/// Polling the future delegates to the wrapped function.
+/// Polling the future delegates to the wrapped function. If the returned future is pinned, then the
+/// captured environment of the wrapped function is also pinned in-place, so as long as the closure
+/// does not move out of its captures it can soundly create pinned references to them.
 ///
 /// # Examples
 ///
@@ -41,7 +43,7 @@ pub struct PollFn<F> {
 }
 
 #[stable(feature = "future_poll_fn", since = "1.64.0")]
-impl<F> Unpin for PollFn<F> {}
+impl<F: Unpin> Unpin for PollFn<F> {}
 
 #[stable(feature = "future_poll_fn", since = "1.64.0")]
 impl<F> fmt::Debug for PollFn<F> {
@@ -57,7 +59,8 @@ where
 {
     type Output = T;
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<T> {
-        (&mut self.f)(cx)
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<T> {
+        // SAFETY: We are not moving out of the pinned field.
+        (unsafe { &mut self.get_unchecked_mut().f })(cx)
     }
 }

